### ISSUE TITLE
Compiling C with Zig 0.8.0 fails on macOS Mojave

### ISSUE THREAD

**Issue by carlocab** [ziglang/zig#8999]
We're currently trying to update Homebrew's Zig to 0.8.0. This builds fine, but the test fails at building a `Hello, world!` program in C, but only on Mojave (macOS 10.14). Everything works fine on Catalina (10.15) and Big Sur (11, both Intel and ARM).

It returns a number of errors of the form:
```
==> /usr/local/Cellar/zig/0.8.0/bin/zig cc hello.c -o hello
In file included from hello.c:1:
In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/stdio.h:64:
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/_stdio.h:93:16: warning: pointer is missing a nullability type specifier (_Nonnull, _Nullable, or _Null_unspecified) [-Wnullability-completeness]
        unsigned char   *_base;
                        ^
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/_stdio.h:93:16: note: insert '_Nullable' if the pointer may be null
        unsigned char   *_base;
                        ^
                          _Nullable 
```

We build Zig with the following CMake invocation:
```
cmake . -DZIG_STATIC_LLVM=ON -DCMAKE_INSTALL_PREFIX=$prefix -DCMAKE_INSTALL_LIBDIR=lib -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev -DBUILD_TESTING=OFF -DCMAKE_OSX_SYSROOT=/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk
```

The version bump PR is https://github.com/Homebrew/homebrew-core/pull/78774, and CI logs are available at https://github.com/Homebrew/homebrew-core/pull/78774/checks?check_run_id=2753820571. The failures start at https://github.com/Homebrew/homebrew-core/pull/78774/checks?check_run_id=2753820571#step:7:55.

Assistance on this would be appreciated.

**Comment by kubkon** [comment]
Thanks for the report @carlocab. The error is actually the below:

```
In file included from hello.c:1:
/Library/Developer/CommandLineTools/SDKs/MacOSX10.14.sdk/usr/include/stdio.h:224:5: error: 'TARGET_OS_EMBEDDED' is not defined, evaluates to 0 [-Werror,-Wundef-prefix=TARGET_OS_]
#if TARGET_OS_EMBEDDED
```

**Comment by kubkon** [comment]
Summoning @mikdusan: would you have any ideas about this? I'm pretty clueless here TBH.

**Comment by kubkon** [comment]
Might be related: https://reviews.llvm.org/D83250

**Comment by mikdusan** [comment]
can we verify a few things; post the output to these commands from BOTH mohave and catalina:
- `zig build-obj --show-builtin`
- `touch empty.c; zig cc -c empty.c -v`

**Comment by kubkon** [comment]
> can we verify a few things; post the output to these commands from BOTH mohave and catalina:
> 
>     * `zig build-obj --show-builtin`
> 
>     * `touch empty.c; zig cc -c empty.c -v`

I don't have access to Mojave machine. @carlocab would you be able to help us out here?

**Comment by carlocab** [comment]
I don't have one either, I'm afraid. I'll see what I can do.

**Comment by mikdusan** [comment]
@carlocab are you on discord or irc? We can probably save time meeting there.

**Comment by mikdusan** [comment]
try this in a temporary window on the failing machine:
- remove homebrew/cellar/whatever from PATH
- assert `which cc` is system compiler from `/usr/bin` and `cc --version` shows "Apple ..." on the first line
- run the same `/usr/local/Cellar/zig/0.8.0/bin/zig cc hello.c -o hello`
- if it fails, nuke `zig-cache` and `.cache/zig` and try the above command again

**Comment by canadaduane** [comment]
> can we verify a few things; post the output to these commands from BOTH mohave and catalina:

I'm on Mojave (10.14.6 ): 

>     * `zig build-obj --show-builtin`

```
const std = @import("std");
/// Zig version. When writing code that supports multiple versions of Zig, prefer
/// feature detection (i.e. with `@hasDecl` or `@hasField`) over version checks.
pub const zig_version = std.SemanticVersion.parse("0.8.0") catch unreachable;
/// Temporary until self-hosted is feature complete.
pub const zig_is_stage2 = true;
/// Temporary until self-hosted supports the `cpu.arch` value.
pub const stage2_arch: std.Target.Cpu.Arch = .x86_64;

pub const output_mode = std.builtin.OutputMode.Obj;
pub const link_mode = std.builtin.LinkMode.Static;
pub const is_test = false;
pub const single_threaded = false;
pub const abi = std.Target.Abi.gnu;
pub const cpu: std.Target.Cpu = .{
    .arch = .x86_64,
    .model = &std.Target.x86.cpu.haswell,
    .features = std.Target.x86.featureSet(&[_]std.Target.x86.Feature{
        .@"64bit",
        .aes,
        .avx,
        .avx2,
        .bmi,
        .bmi2,
        .cmov,
        .cx16,
        .cx8,
        .ermsb,
        .f16c,
        .false_deps_lzcnt_tzcnt,
        .false_deps_popcnt,
        .fast_15bytenop,
        .fast_scalar_fsqrt,
        .fast_shld_rotate,
        .fast_variable_shuffle,
        .fma,
        .fsgsbase,
        .fxsr,
        .idivq_to_divl,
        .invpcid,
        .lzcnt,
        .macrofusion,
        .mmx,
        .movbe,
        .nopl,
        .pclmul,
        .popcnt,
        .rdrnd,
        .sahf,
        .slow_3ops_lea,
        .sse,
        .sse2,
        .sse3,
        .sse4_1,
        .sse4_2,
        .ssse3,
        .vzeroupper,
        .x87,
        .xsave,
        .xsaveopt,
    }),
};
pub const os = std.Target.Os{
    .tag = .macos,
    .version_range = .{ .semver = .{
        .min = .{
            .major = 10,
            .minor = 14,
            .patch = 6,
        },
        .max = .{
            .major = 10,
            .minor = 14,
            .patch = 6,
        },
    }},
};
pub const target = std.Target{
    .cpu = cpu,
    .os = os,
    .abi = abi,
};
pub const object_format = std.Target.ObjectFormat.macho;
pub const mode = std.builtin.Mode.Debug;
pub const link_libc = true;
pub const link_libcpp = false;
pub const have_error_return_tracing = true;
pub const valgrind_support = true;
pub const position_independent_code = true;
pub const position_independent_executable = true;
pub const strip_debug_info = false;
pub const code_model = std.builtin.CodeModel.default;
```

>     * `touch empty.c; zig cc -c empty.c -v`

```
clang version 12.0.1 (git@github.com:ziglang/zig-bootstrap.git 8cc2870e09320a390cafe4e23624e8ed40bd363c)
Target: x86_64-unknown-macosx-gnu
Thread model: posix
InstalledDir: /usr/bin
 (in-process)
 "/Users/duane/.zig/zig" -cc1 -triple x86_64-unknown-macosx10.14.6-gnu -Wundef-prefix=TARGET_OS_ -Werror=undef-prefix -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj --mrelax-relocations -disable-free -disable-llvm-verifier -discard-value-names -main-file-name empty.c -mrelocation-model pic -pic-level 2 -mframe-pointer=none -fno-rounding-math -munwind-tables -fcompatibility-qualified-id-block-type-checking -fvisibility-inlines-hidden-static-local-var -target-cpu penryn -tune-cpu generic -debugger-tuning=lldb -target-linker-version 2.31.1 -v -nostdsysteminc -nobuiltininc -resource-dir /Users/duane/lib/clang/12.0.1 -dependency-file /Users/duane/.cache/zig/tmp/40192370b41654bc-empty.o.d -MT /Users/duane/.cache/zig/tmp/40192370b41654bc-empty.o -sys-header-deps -MV -isystem /Users/duane/.zig/lib/include -isystem /Users/duane/.zig/lib/libc/include/x86_64-macos-gnu -isystem /Users/duane/.zig/lib/libc/include/generic-darwin -isystem /Users/duane/.zig/lib/libc/include/x86_64-macos-any -isystem /Users/duane/.zig/lib/libc/include/any-macos-any -isystem /usr/include -isystem /usr/local/include -iframework /Library/Frameworks -iframework /System/Library/Frameworks -D _DEBUG -Og -fdebug-compilation-dir /Users/duane/Projects/zig-spin -ferror-limit 19 -fsanitize=alignment,array-bounds,bool,builtin,enum,float-cast-overflow,function,integer-divide-by-zero,nonnull-attribute,null,object-size,pointer-overflow,return,returns-nonnull-attribute,shift-base,shift-exponent,signed-integer-overflow,unreachable,vla-bound -fsanitize-trap=alignment,array-bounds,bool,builtin,enum,float-cast-overflow,function,integer-divide-by-zero,nonnull-attribute,null,object-size,pointer-overflow,return,returns-nonnull-attribute,shift-base,shift-exponent,signed-integer-overflow,unreachable,vla-bound -stack-protector 2 -stack-protector-buffer-size 4 -fblocks -fencode-extended-block-signature -fregister-global-dtors-with-atexit -fgnuc-version=4.2.1 -fmax-type-align=16 -fcolor-diagnostics -fno-spell-checking -target-cpu haswell -target-feature -16bit-mode -target-feature -32bit-mode -target-feature -3dnow -target-feature -3dnowa -target-feature +64bit -target-feature -adx -target-feature +aes -target-feature -amx-bf16 -target-feature -amx-int8 -target-feature -amx-tile -target-feature +avx -target-feature +avx2 -target-feature -avx512bf16 -target-feature -avx512bitalg -target-feature -avx512bw -target-feature -avx512cd -target-feature -avx512dq -target-feature -avx512er -target-feature -avx512f -target-feature -avx512ifma -target-feature -avx512pf -target-feature -avx512vbmi -target-feature -avx512vbmi2 -target-feature -avx512vl -target-feature -avx512vnni -target-feature -avx512vp2intersect -target-feature -avx512vpopcntdq -target-feature -avxvnni -target-feature +bmi -target-feature +bmi2 -target-feature -branchfusion -target-feature -cldemote -target-feature -clflushopt -target-feature -clwb -target-feature -clzero -target-feature +cmov -target-feature +cx16 -target-feature +cx8 -target-feature -enqcmd -target-feature +ermsb -target-feature +f16c -target-feature +false-deps-lzcnt-tzcnt -target-feature +false-deps-popcnt -target-feature -fast-11bytenop -target-feature +fast-15bytenop -target-feature -fast-7bytenop -target-feature -fast-bextr -target-feature -fast-gather -target-feature -fast-hops -target-feature -fast-lzcnt -target-feature +fast-scalar-fsqrt -target-feature -fast-scalar-shift-masks -target-feature +fast-shld-rotate -target-feature +fast-variable-shuffle -target-feature -fast-vector-fsqrt -target-feature -fast-vector-shift-masks -target-feature +fma -target-feature -fma4 -target-feature +fsgsbase -target-feature -fsrm -target-feature +fxsr -target-feature -gfni -target-feature -hreset -target-feature -idivl-to-divb -target-feature +idivq-to-divl -target-feature +invpcid -target-feature -kl -target-feature -lea-sp -target-feature -lea-uses-ag -target-feature -lvi-cfi -target-feature -lvi-load-hardening -target-feature -lwp -target-feature +lzcnt -target-feature +macrofusion -target-feature +mmx -target-feature +movbe -target-feature -movdir64b -target-feature -movdiri -target-feature -mwaitx -target-feature +nopl -target-feature -pad-short-functions -target-feature +pclmul -target-feature -pconfig -target-feature -pku -target-feature +popcnt -target-feature -prefer-128-bit -target-feature -prefer-256-bit -target-feature -prefer-mask-registers -target-feature -prefetchwt1 -target-feature -prfchw -target-feature -ptwrite -target-feature -rdpid -target-feature +rdrnd -target-feature -rdseed -target-feature -retpoline -target-feature -retpoline-external-thunk -target-feature -retpoline-indirect-branches -target-feature -retpoline-indirect-calls -target-feature -rtm -target-feature +sahf -target-feature -serialize -target-feature -seses -target-feature -sgx -target-feature -sha -target-feature -shstk -target-feature +slow-3ops-lea -target-feature -slow-incdec -target-feature -slow-lea -target-feature -slow-pmaddwd -target-feature -slow-pmulld -target-feature -slow-shld -target-feature -slow-two-mem-ops -target-feature -slow-unaligned-mem-16 -target-feature -slow-unaligned-mem-32 -target-feature -soft-float -target-feature +sse -target-feature +sse2 -target-feature +sse3 -target-feature +sse4.1 -target-feature +sse4.2 -target-feature -sse4a -target-feature -sse-unaligned-mem -target-feature +ssse3 -target-feature -tbm -target-feature -tsxldtrk -target-feature -uintr -target-feature -use-aa -target-feature -use-glm-div-sqrt-costs -target-feature -vaes -target-feature -vpclmulqdq -target-feature +vzeroupper -target-feature -waitpkg -target-feature -wbnoinvd -target-feature -widekl -target-feature +x87 -target-feature -xop -target-feature +xsave -target-feature -xsavec -target-feature +xsaveopt -target-feature -xsaves -o /Users/duane/.cache/zig/tmp/40192370b41654bc-empty.o -x c empty.c
clang -cc1 version 12.0.1 based upon LLVM 12.0.1 default target x86_64-darwin18.7.0
ignoring nonexistent directory "/Users/duane/.zig/lib/libc/include/generic-darwin"
ignoring nonexistent directory "/Users/duane/.zig/lib/libc/include/x86_64-macos-any"
ignoring nonexistent directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 /Users/duane/.zig/lib/include
 /Users/duane/.zig/lib/libc/include/x86_64-macos-gnu
 /Users/duane/.zig/lib/libc/include/any-macos-any
 /usr/local/include
 /Library/Frameworks (framework directory)
 /System/Library/Frameworks (framework directory)
End of search list.
```

**Comment by mikdusan** [comment]
one quick thing to try is on the machine that is failing, can you temporarily do something like this to make sure `/usr/local/include` is not found? I'd like to rule out interference from that folder.

```sh
mv /usr/local/include /usr/local/include.DISABLED
```

and then repeat issue:
```sh
/usr/local/Cellar/zig/0.8.0/bin/zig cc hello.c -o hello
```

ps. Don't forget to re-enable that folder again.

**Comment by kubkon** [comment]
Just an FYI @carlocab, a fixed landed in master that will be backported to 0.8.1 release.

**Comment by carlocab** [comment]
Great; I've opened Homebrew/homebrew-core#81926 to apply the patch from #9427.
